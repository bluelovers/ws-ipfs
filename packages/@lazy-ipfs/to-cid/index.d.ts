/**
 * Created by user on 2020/5/17.
 */
import JsCID from 'cids';
import { CID as MultiformatsCID } from 'multiformats';
import { ICIDObject, ICIDObjectInput, ICIDValueInput, IRawCIDObject } from '@lazy-ipfs/detect-cid-lib/lib/types';
import { EnumTypeofCID } from '@lazy-ipfs/detect-cid-lib';
import { IBaseNameOrBaseCodec } from '@lazy-ipfs/cid-to-string';
import { IParsePathResult } from '@lazy-ipfs/parse-ipfs-path/lib/types';
export * from '@lazy-ipfs/detect-cid-lib/lib/types';
export { SymbolJsCID as SymbolCID } from '@lazy-ipfs/detect-cid-lib/lib/js-cids';
export type IStaticCID<C extends ICIDObject = ICIDObject> = new (...argv: any[]) => C;
export type IToCIDInputValue = ICIDValueInput | IParsePathResult;
export declare function classCID<C extends ICIDObject = MultiformatsCID>(libCID?: IStaticCID<C> | typeof MultiformatsCID | typeof JsCID | EnumTypeofCID): <T extends IToCIDInputValue>(cidInput: T, libCID?: IStaticCID<C> | EnumTypeofCID) => C;
export declare function isCID<C extends ICIDObject = ICIDObject>(cid: unknown, libCID?: IStaticCID<C> | EnumTypeofCID): cid is C;
export declare function assertRawCIDLike<C extends IRawCIDObject = IRawCIDObject>(cid: unknown): asserts cid is C;
export declare function isRawCIDLike<T extends IRawCIDObject = IRawCIDObject>(cid: any): cid is T;
export declare function toRawCID<R extends IRawCIDObject = IRawCIDObject>(cid: ICIDObjectInput): R;
export declare function toCID<C extends ICIDObject = MultiformatsCID>(cid: IToCIDInputValue, libCID?: IStaticCID<C> | EnumTypeofCID): C;
export declare function strToCidToStr(str: string, base?: IBaseNameOrBaseCodec): string;
export default toCID;
