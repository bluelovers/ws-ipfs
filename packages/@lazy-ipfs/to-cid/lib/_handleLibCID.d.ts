import { EnumTypeofCID, ICIDObject } from '@lazy-ipfs/detect-cid-lib';
import { IStaticCID } from '../index';
import JsCID from 'cids';
import { CID as MultiformatsCID } from 'multiformats';
export declare type ILibCID<C extends ICIDObject = ICIDObject> = IStaticCID<C> | typeof JsCID | Pick<typeof MultiformatsCID, 'parse' | 'decode' | 'asCID'> | IStaticCID<JsCID>;
export declare function _handleLibCID<C extends ILibCID>(libCID: C | EnumTypeofCID, defaultLibCID?: ILibCID): C;
