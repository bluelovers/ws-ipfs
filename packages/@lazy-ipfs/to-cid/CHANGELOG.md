# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.4](https://github.com/bluelovers/ws-ipfs/compare/@lazy-ipfs/to-cid@1.0.3...@lazy-ipfs/to-cid@1.0.4) (2020-06-05)


### Features

* update deps ([a25ebd6](https://github.com/bluelovers/ws-ipfs/commit/a25ebd688ccfd54f164b3ff89cf6cdb2e7f6e478))





## [1.0.3](https://github.com/bluelovers/ws-ipfs/compare/@lazy-ipfs/to-cid@1.0.2...@lazy-ipfs/to-cid@1.0.3) (2020-06-05)

**Note:** Version bump only for package @lazy-ipfs/to-cid





## [1.0.2](https://github.com/bluelovers/ws-ipfs/compare/@lazy-ipfs/to-cid@1.0.1...@lazy-ipfs/to-cid@1.0.2) (2020-05-24)

**Note:** Version bump only for package @lazy-ipfs/to-cid





## 1.0.1 (2020-05-17)


### Features

* @lazy-ipfs/to-cid ([ed8db4a](https://github.com/bluelovers/ws-ipfs/commit/ed8db4abbd46c4c13878223980fd3794dcc882a5))
* no need check multibaseName add test @lazy-ipfs/to-cid ([e6ea250](https://github.com/bluelovers/ws-ipfs/commit/e6ea250cad011f1c0aaf1558df2201e6723d291e))
