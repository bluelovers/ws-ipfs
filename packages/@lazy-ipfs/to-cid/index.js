"use strict";
/**
 * Created by user on 2020/5/17.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.strToCidToStr = exports.toCID = exports.toRawCID = exports.isRawCIDLike = exports.assertRawCIDLike = exports.isCID = exports.classCID = exports.SymbolCID = void 0;
const tslib_1 = require("tslib");
const cids_1 = (0, tslib_1.__importDefault)(require("cids"));
const multiformats_1 = require("multiformats");
const js_cids_1 = require("@lazy-ipfs/detect-cid-lib/lib/js-cids");
const index_1 = (0, tslib_1.__importStar)(require("@lazy-ipfs/detect-cid-lib/index"));
const err_code_1 = (0, tslib_1.__importDefault)(require("err-code"));
const multiformats_2 = (0, tslib_1.__importDefault)(require("./lib/multiformats"));
const js_cids_2 = require("./lib/js-cids");
const index_2 = require("@lazy-ipfs/cid-to-string/index");
(0, tslib_1.__exportStar)(require("@lazy-ipfs/detect-cid-lib/lib/types"), exports);
var js_cids_3 = require("@lazy-ipfs/detect-cid-lib/lib/js-cids");
Object.defineProperty(exports, "SymbolCID", { enumerable: true, get: function () { return js_cids_3.SymbolJsCID; } });
function classCID(libCID) {
    libCID !== null && libCID !== void 0 ? libCID : (libCID = multiformats_1.CID);
    if (libCID === cids_1.default || libCID === "@ipld/js-cid/CID" /* js_cids */) {
        return js_cids_2.toJsCID;
    }
    return multiformats_2.default;
}
exports.classCID = classCID;
function isCID(cid, libCID) {
    const type = (0, index_1.default)(cid);
    // @ts-ignore
    if (libCID === multiformats_1.CID || libCID === "@ipld/js-multiformats/CID" /* js_multiformats */) {
        return type === "@ipld/js-multiformats/CID" /* multiformats_cid */;
    }
    // @ts-ignore
    else if (libCID === cids_1.default || libCID === "@ipld/js-cid/CID" /* js_cids */) {
        return type === "@ipld/js-cid/CID" /* js_cids */;
    }
    return (type === null || type === void 0 ? void 0 : type.length) > 0;
}
exports.isCID = isCID;
function assertRawCIDLike(cid) {
    if (!isRawCIDLike(cid)) {
        throw (0, err_code_1.default)(new TypeError(`Invalid type for CID like`), {
            input: cid,
        });
    }
}
exports.assertRawCIDLike = assertRawCIDLike;
function isRawCIDLike(cid) {
    return (0, index_1.isRawMultiformatsCIDLike)(cid) || (0, js_cids_1.isRawJsCIDLike)(cid);
}
exports.isRawCIDLike = isRawCIDLike;
function toRawCID(cid) {
    assertRawCIDLike(cid);
    if ((0, index_1.isRawMultiformatsCIDLike)(cid)) {
        return (0, index_1.toRawMultiformatsCID)(cid);
    }
    return (0, js_cids_1.toRawJsCID)(cid);
}
exports.toRawCID = toRawCID;
function toCID(cid, libCID) {
    return classCID(libCID)(cid, libCID);
}
exports.toCID = toCID;
function strToCidToStr(str, base) {
    return (0, index_2.cidToString)(toCID(str), base);
}
exports.strToCidToStr = strToCidToStr;
exports.default = toCID;
//# sourceMappingURL=index.js.map