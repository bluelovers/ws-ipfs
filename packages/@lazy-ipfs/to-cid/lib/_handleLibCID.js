"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._handleLibCID = void 0;
const tslib_1 = require("tslib");
const cids_1 = tslib_1.__importDefault(require("cids"));
const multiformats_1 = require("multiformats");
function _handleLibCID(libCID, defaultLibCID) {
    if (libCID === "@ipld/js-cid/CID" /* EnumTypeofCID.js_cids */) {
        return cids_1.default;
    }
    else if (libCID === "@ipld/js-multiformats/CID" /* EnumTypeofCID.js_multiformats */) {
        return multiformats_1.CID;
    }
    return libCID !== null && libCID !== void 0 ? libCID : defaultLibCID;
}
exports._handleLibCID = _handleLibCID;
//# sourceMappingURL=_handleLibCID.js.map