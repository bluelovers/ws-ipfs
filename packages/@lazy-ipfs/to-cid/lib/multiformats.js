"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toMultiformatsCID = void 0;
const tslib_1 = require("tslib");
const err_code_1 = (0, tslib_1.__importDefault)(require("err-code"));
const multiformats_1 = require("multiformats");
const js_cids_1 = require("@lazy-ipfs/detect-cid-lib/lib/js-cids");
const js_multiformats_1 = require("@lazy-ipfs/detect-cid-lib/lib/js-multiformats");
const util_1 = require("@lazy-ipfs/detect-cid-lib/lib/util");
const parsePath_1 = require("@lazy-ipfs/parse-ipfs-path/lib/parsePath");
const _handleLibCID_1 = require("./_handleLibCID");
function toMultiformatsCID(cidInput, libCID) {
    libCID = (0, _handleLibCID_1._handleLibCID)(libCID, multiformats_1.CID);
    if (typeof cidInput === 'string') {
        return libCID.parse((0, parsePath_1.parsePath)(cidInput, {
            unsafeReturn: true,
            noThrow: true,
        }).hash);
    }
    else if ((0, util_1._isArrayLike)(cidInput)) {
        return libCID.decode(cidInput);
    }
    let cid = libCID.asCID(cidInput);
    if ((typeof cid === 'undefined' || cid === null)) {
        if ((0, js_cids_1.isRawJsCIDLike)(cidInput)) {
            cid = libCID.asCID((0, js_cids_1.toRawJsCIDFake)(cidInput));
        }
        else if ((0, js_multiformats_1.isRawMultiformatsCIDLike)(cidInput)) {
            cid = libCID.asCID((0, js_multiformats_1.toRawMultiformatsCIDFake)(cidInput));
        }
        else if ((0, parsePath_1.isParsePathResultLoose)(cidInput)) {
            cid = toMultiformatsCID(cidInput.hash);
        }
    }
    if (!cid) {
        throw (0, err_code_1.default)(new TypeError(`Invalid type for convert to MultiformatsCID`), {
            input: cidInput,
        });
    }
    return cid;
}
exports.toMultiformatsCID = toMultiformatsCID;
exports.default = toMultiformatsCID;
//# sourceMappingURL=multiformats.js.map