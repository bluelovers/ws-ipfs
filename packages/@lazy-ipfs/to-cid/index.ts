/**
 * Created by user on 2020/5/17.
 */

import JsCID from 'cids';
import { CID as MultiformatsCID } from 'multiformats';
import { isRawJsCIDLike, toRawJsCID } from '@lazy-ipfs/detect-cid-lib/lib/js-cids';
import { ICIDObject, ICIDObjectInput, ICIDValueInput, IRawCIDObject } from '@lazy-ipfs/detect-cid-lib/lib/types';
import typeofCID, {
	EnumTypeofCID,
	isRawMultiformatsCIDLike,
	toRawMultiformatsCID,
} from '@lazy-ipfs/detect-cid-lib/index';
import err_code from 'err-code';
import toMultiformatsCID from './lib/multiformats';
import { toJsCID } from './lib/js-cids';
import { cidToString, IBaseNameOrBaseCodec } from '@lazy-ipfs/cid-to-string/index';

export * from '@lazy-ipfs/detect-cid-lib/lib/types';

export { SymbolJsCID as SymbolCID } from '@lazy-ipfs/detect-cid-lib/lib/js-cids';

export type IStaticCID<C extends ICIDObject = ICIDObject> = new(...argv: any[]) => C;

export function classCID<C extends ICIDObject = MultiformatsCID>(libCID?: IStaticCID<C> | typeof MultiformatsCID | typeof JsCID | EnumTypeofCID): <T extends ICIDValueInput>(cidInput: T, libCID?: IStaticCID<C> | EnumTypeofCID) => C
{
	libCID ??= MultiformatsCID;

	if (libCID === JsCID || libCID === EnumTypeofCID.js_cids)
	{
		return toJsCID as any
	}

	return toMultiformatsCID as any
}

export function isCID<C extends ICIDObject = ICIDObject>(cid: unknown, libCID?: IStaticCID<C> | EnumTypeofCID): cid is C
{
	const type = typeofCID(cid);

	// @ts-ignore
	if (libCID === MultiformatsCID || libCID === EnumTypeofCID.js_multiformats)
	{
		return type === EnumTypeofCID.multiformats_cid
	}
	// @ts-ignore
	else if (libCID === JsCID || libCID === EnumTypeofCID.js_cids)
	{
		return type === EnumTypeofCID.js_cids
	}

	return type?.length > 0
}

export function assertRawCIDLike<C extends IRawCIDObject = IRawCIDObject>(cid: unknown): asserts cid is C
{
	if (!isRawCIDLike(cid))
	{
		throw err_code(new TypeError(`Invalid type for CID like`), {
			input: cid,
		})
	}
}

export function isRawCIDLike<T extends IRawCIDObject = IRawCIDObject>(cid: any): cid is T
{
	return isRawMultiformatsCIDLike(cid) || isRawJsCIDLike(cid)
}

export function toRawCID<R extends IRawCIDObject = IRawCIDObject>(cid: ICIDObjectInput): R
{
	assertRawCIDLike(cid);

	if (isRawMultiformatsCIDLike(cid))
	{
		return toRawMultiformatsCID(cid) as R
	}

	return toRawJsCID(cid) as R
}

export function toCID<C extends ICIDObject = ICIDObject>(cid: any, libCID?: IStaticCID<C> | EnumTypeofCID): C
{
	return classCID(libCID)(cid, libCID)
}

export function strToCidToStr(str: string, base?: IBaseNameOrBaseCodec)
{
	return cidToString(toCID(str), base)
}

export default toCID;
