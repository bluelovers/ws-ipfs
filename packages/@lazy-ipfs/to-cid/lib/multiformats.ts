import err_code from 'err-code';
import { CID as MultiformatsCID } from 'multiformats'
import { isRawJsCIDLike, SymbolJsCID, toRawJsCIDFake } from '@lazy-ipfs/detect-cid-lib/lib/js-cids';
import { ICIDValueInput, ICIDValueOrRaw } from '@lazy-ipfs/detect-cid-lib/lib/types';
import { IRawMultiformatsCIDFake, isRawMultiformatsCIDLike, toRawMultiformatsCIDFake } from '@lazy-ipfs/detect-cid-lib/lib/js-multiformats';
import { _isArrayLike } from '@lazy-ipfs/detect-cid-lib/lib/util';
import { parsePath } from '@lazy-ipfs/parse-ipfs-path/lib/parsePath';
import { _handleLibCID } from './_handleLibCID';
import { EnumTypeofCID } from '@lazy-ipfs/detect-cid-lib';

export function toMultiformatsCID<T extends ICIDValueInput, C extends MultiformatsCID = MultiformatsCID>(cidInput: T, libCID?: Pick<typeof MultiformatsCID, 'parse' | 'decode' | 'asCID'> | EnumTypeofCID): C
{
	libCID = _handleLibCID(libCID, MultiformatsCID);

	if (typeof cidInput === 'string')
	{
		return libCID.parse(parsePath(cidInput, {
			unsafeReturn: true,
			noThrow: true,
		}).hash) as any
	}
	else if (_isArrayLike(cidInput))
	{
		return libCID.decode(cidInput) as any
	}

	let cid = libCID.asCID(cidInput);

	if ((typeof cid === 'undefined' || cid === null))
	{
		if (isRawJsCIDLike(cidInput))
		{
			cid = libCID.asCID(toRawJsCIDFake(cidInput));
		}
		else if (isRawMultiformatsCIDLike(cidInput))
		{
			cid = libCID.asCID(toRawMultiformatsCIDFake(cidInput))
		}
	}

	if (!cid)
	{
		throw err_code(new TypeError(`Invalid type for convert to MultiformatsCID`), {
			input: cidInput,
		})
	}

	return cid as any
}

export default toMultiformatsCID
