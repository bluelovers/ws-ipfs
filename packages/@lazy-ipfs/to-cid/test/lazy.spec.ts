import { toCID } from '@lazy-ipfs/to-cid';
import { handleCID } from 'fetch-ipfs/util';
import { cidToString } from '@lazy-ipfs/cid-to-string';
import { parsePath } from '@lazy-ipfs/parse-ipfs-path/lib/parsePath';
import JsCID from 'cids';
import { EnumTypeofCID } from '@lazy-ipfs/detect-cid-lib';
import { CID as MultiformatsCID } from 'multiformats';

const cidStr = 'QmTWsPwz1wrKSgpiKa5tyC3rWkWh3Kym6ikCSmwiGPsfqd';

describe(`describe`, () =>
{

	test(`url`, () =>
	{

		_check(handleCID(`https://ipfs.io/ipfs/${cidStr}?filename=wenku8_1508.epub`, true))

	});

	test(`/ipfs/hash`, () =>
	{

		_check(`ipfs/${cidStr}`)

	});

	test(`ParsePathResult`, () =>
	{

		_check(parsePath(cidStr, {
			unsafeReturn: true,
			noThrow: true,
		}))

	});

	test(EnumTypeofCID.js_cids, () =>
	{
		const oldCid = new JsCID(cidStr)

		_check(oldCid)

	});

	test(EnumTypeofCID.js_multiformats, () =>
	{
		const newCid = MultiformatsCID.parse(cidStr)

		_check(newCid)

	});

	test('cidStr', () =>
	{
		_check(cidStr)

	});

})

function _check(input)
{
	let actual = toCID(input);
	let expected = cidStr;

	expect(cidToString(actual)).toStrictEqual(expected);

	expect(actual).toMatchSnapshot();
	expect(cidToString(actual)).toMatchSnapshot();
}
