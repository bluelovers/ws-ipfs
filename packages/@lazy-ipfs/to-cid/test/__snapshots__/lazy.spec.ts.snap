// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe /ipfs/hash 1`] = `
{
  "code": 112,
  "hash": Uint8Array [
    18,
    32,
    76,
    232,
    150,
    127,
    149,
    124,
    163,
    248,
    248,
    180,
    37,
    130,
    164,
    234,
    122,
    178,
    127,
    106,
    233,
    159,
    188,
    136,
    165,
    115,
    201,
    56,
    146,
    24,
    205,
    187,
    255,
    44,
  ],
  "version": 0,
}
`;

exports[`describe /ipfs/hash 2`] = `"QmTWsPwz1wrKSgpiKa5tyC3rWkWh3Kym6ikCSmwiGPsfqd"`;

exports[`describe @ipld/js-cid/CID 1`] = `
{
  "code": 112,
  "hash": Uint8Array [
    18,
    32,
    76,
    232,
    150,
    127,
    149,
    124,
    163,
    248,
    248,
    180,
    37,
    130,
    164,
    234,
    122,
    178,
    127,
    106,
    233,
    159,
    188,
    136,
    165,
    115,
    201,
    56,
    146,
    24,
    205,
    187,
    255,
    44,
  ],
  "version": 0,
}
`;

exports[`describe @ipld/js-cid/CID 2`] = `"QmTWsPwz1wrKSgpiKa5tyC3rWkWh3Kym6ikCSmwiGPsfqd"`;

exports[`describe @ipld/js-multiformats/CID 1`] = `
{
  "code": 112,
  "hash": Uint8Array [
    18,
    32,
    76,
    232,
    150,
    127,
    149,
    124,
    163,
    248,
    248,
    180,
    37,
    130,
    164,
    234,
    122,
    178,
    127,
    106,
    233,
    159,
    188,
    136,
    165,
    115,
    201,
    56,
    146,
    24,
    205,
    187,
    255,
    44,
  ],
  "version": 0,
}
`;

exports[`describe @ipld/js-multiformats/CID 2`] = `"QmTWsPwz1wrKSgpiKa5tyC3rWkWh3Kym6ikCSmwiGPsfqd"`;

exports[`describe ParsePathResult 1`] = `
{
  "code": 112,
  "hash": Uint8Array [
    18,
    32,
    76,
    232,
    150,
    127,
    149,
    124,
    163,
    248,
    248,
    180,
    37,
    130,
    164,
    234,
    122,
    178,
    127,
    106,
    233,
    159,
    188,
    136,
    165,
    115,
    201,
    56,
    146,
    24,
    205,
    187,
    255,
    44,
  ],
  "version": 0,
}
`;

exports[`describe ParsePathResult 2`] = `"QmTWsPwz1wrKSgpiKa5tyC3rWkWh3Kym6ikCSmwiGPsfqd"`;

exports[`describe cidStr 1`] = `
{
  "code": 112,
  "hash": Uint8Array [
    18,
    32,
    76,
    232,
    150,
    127,
    149,
    124,
    163,
    248,
    248,
    180,
    37,
    130,
    164,
    234,
    122,
    178,
    127,
    106,
    233,
    159,
    188,
    136,
    165,
    115,
    201,
    56,
    146,
    24,
    205,
    187,
    255,
    44,
  ],
  "version": 0,
}
`;

exports[`describe cidStr 2`] = `"QmTWsPwz1wrKSgpiKa5tyC3rWkWh3Kym6ikCSmwiGPsfqd"`;

exports[`describe url 1`] = `
{
  "code": 112,
  "hash": Uint8Array [
    18,
    32,
    76,
    232,
    150,
    127,
    149,
    124,
    163,
    248,
    248,
    180,
    37,
    130,
    164,
    234,
    122,
    178,
    127,
    106,
    233,
    159,
    188,
    136,
    165,
    115,
    201,
    56,
    146,
    24,
    205,
    187,
    255,
    44,
  ],
  "version": 0,
}
`;

exports[`describe url 2`] = `"QmTWsPwz1wrKSgpiKa5tyC3rWkWh3Kym6ikCSmwiGPsfqd"`;
